### ISSUE TITLE
ERROR:asyncio:Future exception was never retrieved from AdHocSignal.AUTO_FUTURE

### ISSUE THREAD

**Issue by pszafer** [horazont/aioxmpp#358]
Hi,
I'm still trying to fix aioxmpp error.
Right now I copied sendxmpp.py example file, changed it according to bosch xmpp and it's not working same as my lib.

In the end I got errors:
```
2021-01-19 09:12:31,396:DEBUG:aioxmpp.node.PresenceManagedClient.on_stopped:connecting <Future pending> with mode <bound method AdHocSignal.AUTO_FUTURE of <class 'aioxmpp.callbacks.AdHocSignal'>>
2021-01-19 09:12:31,396:DEBUG:aioxmpp.node.PresenceManagedClient.on_failure:connecting <Future pending> with mode <bound method AdHocSignal.AUTO_FUTURE of <class 'aioxmpp.callbacks.AdHocSignal'>>
2021-01-19 09:12:31,397:DEBUG:aioxmpp.node.PresenceManagedClient:stopping main task of <aioxmpp.node.PresenceManagedClient object at 0x7f6471e70bb0>
Stack (most recent call last):
  File "$HOME/bosch-thermostat-client-python/venv/bin/bosch_scan", line 33, in <module>
    sys.exit(load_entry_point('bosch-thermostat-client', 'console_scripts', 'bosch_scan')())
  File "$HOME/bosch-thermostat-client-python/venv/lib/python3.9/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "$HOME/bosch-thermostat-client-python/venv/lib/python3.9/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "$HOME/bosch-thermostat-client-python/venv/lib/python3.9/site-packages/click/core.py", line 1236, in invoke
    return Command.invoke(self, ctx)
  File "$HOME/bosch-thermostat-client-python/venv/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "$HOME/bosch-thermostat-client-python/venv/lib/python3.9/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "$HOME/bosch-thermostat-client-python/venv/lib/python3.9/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "$HOME/bosch-thermostat-client-python/bosch_thermostat_client/bosch_rawscan_cli.py", line 18, in wrapper
    return asyncio.run(f(*args, **kwargs))
  File "/usr/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.9/asyncio/base_events.py", line 629, in run_until_complete
    self.run_forever()
  File "/usr/lib/python3.9/asyncio/base_events.py", line 596, in run_forever
    self._run_once()
  File "/usr/lib/python3.9/asyncio/base_events.py", line 1890, in _run_once
    handle._run()
  File "/usr/lib/python3.9/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "$HOME/bosch-thermostat-client-python/bosch_thermostat_client/bosch_rawscan_cli.py", line 63, in cli
    if await gateway.check_connection():
  File "$HOME/bosch-thermostat-client-python/bosch_thermostat_client/gateway/base_gateway.py", line 234, in check_connection
    await self.initialize()
  File "$HOME/bosch-thermostat-client-python/bosch_thermostat_client/gateway/base_gateway.py", line 61, in initialize
    await self._update_info(initial_db.get(GATEWAY))
  File "$HOME/bosch-thermostat-client-python/bosch_thermostat_client/gateway/ivt_gateway.py", line 57, in _update_info
    response = await self._connector.get(uri)
  File "$HOME/bosch-thermostat-client-python/bosch_thermostat_client/connectors/xmpp.py", line 115, in get
    raise DeviceException(f"Error requesting data from {path}")
  File "$HOME/bosch-thermostat-client-python/venv/lib/python3.9/site-packages/aioxmpp/node.py", line 1672, in __aexit__
    self._client.stop()
  File "$HOME/bosch-thermostat-client-python/venv/lib/python3.9/site-packages/aioxmpp/node.py", line 1057, in stop
    self.logger.debug("stopping main task of %r", self, stack_info=True)
2021-01-19 09:12:31,397:INFO:aioxmpp.node.PresenceManagedClient:client shutting down (on request)
2021-01-19 09:12:31,397:DEBUG:aioopenssl.trace.fd=7:_write_ready: nothing more to write, removing writer
2021-01-19 09:12:31,397:DEBUG:aioxmpp.StanzaStream:sending stop signal to task
2021-01-19 09:12:31,397:ERROR:asyncio:Future exception was never retrieved
future: <Future finished exception=ConnectionError('stream shut down')>
ConnectionError: stream shut down
2021-01-19 09:12:31,397:ERROR:asyncio:Future exception was never retrieved
future: <Future finished exception=ConnectionError('stream shut down')>
ConnectionError: stream shut down
```

Full error and source file I uploaded as gist: https://gist.github.com/pszafer/0d0b7a745ec4caa02254d8b8ad2cf070

**Comment by pszafer** [comment]
Hi,
I found that this `set_exception` is not handled.
https://github.com/horazont/aioxmpp/blob/beaa0062237619fc6d23757d3f55b2e05707f8f2/aioxmpp/protocol.py#L397

how to catch it? or is it something lib should do?

**Comment by horazont** [comment]
@pszafer Sorry for the late reply, I’m in the middle of a tricky server migration.

So the stacktrace you have shown does not indicate any aioxmpp problem. Let me go through it step by step:

1. ```
   Stack (most recent call last):
     File "$HOME/bosch-thermostat-client-python/venv/bin/bosch_scan", line 33, in <module>
       sys.exit(load_entry_point('bosch-thermostat-client', 'console_scripts', 'bosch_scan')())
     File "$HOME/bosch-thermostat-client-python/venv/lib/python3.9/site-packages/click/core.py", line 829, in __call__
       return self.main(*args, **kwargs)
     File "$HOME/bosch-thermostat-client-python/venv/lib/python3.9/site-packages/click/core.py", line 782, in main
       rv = self.invoke(ctx)
     File "$HOME/bosch-thermostat-client-python/venv/lib/python3.9/site-packages/click/core.py", line 1236, in invoke
       return Command.invoke(self, ctx)
     File "$HOME/bosch-thermostat-client-python/venv/lib/python3.9/site-packages/click/core.py", line 1066, in invoke
       return ctx.invoke(self.callback, **ctx.params)
     File "$HOME/bosch-thermostat-client-python/venv/lib/python3.9/site-packages/click/core.py", line 610, in invoke
       return callback(*args, **kwargs)
     File "$HOME/bosch-thermostat-client-python/venv/lib/python3.9/site-packages/click/decorators.py", line 21, in new_func
       return f(get_current_context(), *args, **kwargs)
     File "$HOME/bosch-thermostat-client-python/bosch_thermostat_client/bosch_rawscan_cli.py", line 18, in wrapper
       return asyncio.run(f(*args, **kwargs))
     File "/usr/lib/python3.9/asyncio/runners.py", line 44, in run
       return loop.run_until_complete(main)
     File "/usr/lib/python3.9/asyncio/base_events.py", line 629, in run_until_complete
       self.run_forever()
     File "/usr/lib/python3.9/asyncio/base_events.py", line 596, in run_forever
       self._run_once()
     File "/usr/lib/python3.9/asyncio/base_events.py", line 1890, in _run_once
       handle._run()
     File "/usr/lib/python3.9/asyncio/events.py", line 80, in _run
       self._context.run(self._callback, *self._args)
     File "$HOME/bosch-thermostat-client-python/bosch_thermostat_client/bosch_rawscan_cli.py", line 63, in cli
       if await gateway.check_connection():
     File "$HOME/bosch-thermostat-client-python/bosch_thermostat_client/gateway/base_gateway.py", line 234, in check_connection
       await self.initialize()
     File "$HOME/bosch-thermostat-client-python/bosch_thermostat_client/gateway/base_gateway.py", line 61, in initialize
       await self._update_info(initial_db.get(GATEWAY))
     File "$HOME/bosch-thermostat-client-python/bosch_thermostat_client/gateway/ivt_gateway.py", line 57, in _update_info
       response = await self._connector.get(uri)
     File "$HOME/bosch-thermostat-client-python/bosch_thermostat_client/connectors/xmpp.py", line 115, in get
       raise DeviceException(f"Error requesting data from {path}")
     File "$HOME/bosch-thermostat-client-python/venv/lib/python3.9/site-packages/aioxmpp/node.py", line 1672, in __aexit__
       self._client.stop()
     File "$HOME/bosch-thermostat-client-python/venv/lib/python3.9/site-packages/aioxmpp/node.py", line 1057, in stop
       self.logger.debug("stopping main task of %r", self, stack_info=True)
   ```

   This is not a problem at all in aioxmpp, only a hint (hence the DEBUG level). It indicates that the stream is being closed, because the `async with` block is being exited, in your case with an exception (``raise DeviceException(f"Error requesting data from {path}")``). What causes that exception I don’t know, because it’s not in aioxmpp code.

2. ```
   2021-01-19 09:12:31,397:ERROR:asyncio:Future exception was never retrieved
   future: <Future finished exception=ConnectionError('stream shut down')>
   ConnectionError: stream shut down
   ```

   This is more or less a known thing, but it doesn’t impact the operation of aioxmpp at all. It is a red hering. It is unrelated to the problem you have.

I don’t know what the actual problem is and it would be helpful if you explained it.

**Comment by pszafer** [comment]
Thank you for response,

ad 1. I resolved this already, so this logs we can forget.
ad 2.
```
2021-02-23 19:47:52 ERROR (MainThread) [homeassistant] Error doing job: Future exception was never retrieved
ConnectionError: stream shut down
2021-02-23 19:47:52 ERROR (MainThread) [homeassistant] Error doing job: Future exception was never retrieved
ConnectionError: stream shut down
```

People asking me about this error as it appears on restart of Home Assistant and I wanted to know if there is something I could to fix/hide this message.

**Comment by horazont** [comment]
@pszafer If you could figure out for me the stacktrace where that future is created, that’d be great. I’m a bit short on time this week to find that, but I’ve spotted it in a variety of aioxmpp applications which makes me think that it’s a more general thing.

**Comment by horazont** [comment]
@pszafer Nevermind, I dug into it. Can you try if #359 fixes it for you? Otherwise I guess I’ll have to make a release on a hunch :)

**Comment by pszafer** [comment]
Thanks @horazont I'll let you know tomorrow!

**Comment by pszafer** [comment]
This PR https://github.com/horazont/aioxmpp/pull/359 fixes this issue.

**Comment by horazont** [comment]
@pszafer Thanks for the feedback. I’ll try to fix the remaining issues with the fix by the end of the week and cut a 0.12.1 release out of it!

**Comment by horazont** [comment]
I just pushed 0.12.1 to PyPI, which should fix the issue. I’m now working on get 0.12.1 into Debian so that it’ll be included in bullseye. If the issue is not fixed, please report back ASAP.
